src/Filesystem/File.php

```diff
@@ -447,7 +474,18 @@ protected function setUrl(string|null $url = null): static
 	 */
 	public function url(): string|null
 	{
-		return $this->url;
+		if ($this->url !== null) {
+			return $this->url;
+		}
+
+		// lazily determine the URL from the model object
+		// only if it's needed to avoid breaking custom file::url
+		// components that rely on `$cmsFile->asset()` methods
+		if ($this->model !== null) {
+			return $this->url = $this->model->url();
+		}
+
+		return null;
```

Comment:
Lines 481-488 could be simply

```php
return $this->url = $this->model?->url();
```

If model is null the whole return statement will still be null.